Updated spec constant proposal to use NTTP. (#1)
KhronosGroup/SYCL-Shared: proposals/sycl_modules.md

## Revisions

v 0.10:

* Rework of the specialization constant proposal

>   * Add `specialization _id` and `specialization _constant` definitions

Reviewer: Curious spaces before `_`

Author: that's odd,

anyway, now fixed

   * Returns true if the current module uses specialization constants.
   *
   */
  bool has_specialization_constant() const noexcept;

  /**
   * Returns true if the current module can support specialization constants natively.
   *
   */
  bool native_specialization_constant() const noexcept;

  // Available if module_status::executable
  // Retrieve the value of the specialization constant associated with this module.

>   // Call only valid from the host.

Reviewer: Also that the specialization constant has been previously initialized or it will throw.

>   template<class T, specialization_id<T>& s>

Reviewer: After a second review I have just realized the `&` here, meaning that the id is actually the *address* of the `specialization_id` object...
This changes quite a lot in my understanding of the design...

  specialization_constant<T, s> get_specialization_constant(handler&);

  // Available if module_status::executable
  //              Only from C++17

>   // Set the value of the specialization constant.

Reviewer: ```suggestion
  // Get the value of the specialization constant.
```

Author: Fixed

  // Call only valid from the host.
  template<auto& s>
  specialization_constant<typename std::remove_reference_t<decltype(s)>::type, s>

>     get_specialization_constant(handler&);

Reviewer: While waiting for concepts to be mandatory in SYCL, at least you should specificy in the comment that `s` needs to be an existing specialization identier or it will throw. Also that the specialization constant has been previously initialized.

   */
  bool has_spec_constant() const noexcept;

  /**
   * Returns true if the current module can support specialization constants natively.
   *
   */
  bool native_spec_constant() const noexcept;

  // Available if module_status::executable

>   // Retrive the value of the spec constant associated with this module.

Reviewer: It does not make sense to use abbreviation inside the comments, since comments are not about obfuscation. :-)
spec → specialization constant
Retrive → Retrieve

Author: Good point, I fixed all occurrences of `spec `

  // Call only valid from the host.
  template<class T, spec_id<T>& s>
  spec_constant<T, s> get_spec_constant(handler&);

  // Available if module_status::executable
  //              Only from C++17

>   // Set the value of the spec constant.

Reviewer: specialization

  // Call only valid from the host.
  template<auto& s>
  spec_constant<typename std::remove_reference_t<decltype(s)>::type, s>
    get_spec_constant(handler&);

  // Available if module_status::input

>   // Set the value of the spec constant.

Reviewer: specialization

  // Call only valid from the host.
  template<class T, spec_id<T>& s>
  void set_spec_constant(T);

  // Available if module_status::input
  //              Only from C++17

>   // Set the value of the spec constant.

Reviewer: ```suggestion
  // Set the value of the specialization constant.
```

To retrive the value of a spec constant in a SYCL kernel context, the following API is available.

```cpp
namespace sycl {

template<class T, spec_id<T>& s>
T get_spec_constant(const spec_constant<T, s>&);

}  // namespace sycl
```
If C++17 or later is enabled, then the following overload is available
```cpp
namespace sycl {

template<auto& s>
typename std::remove_reference_t<decltype(s)>::type
get_spec_constant(const spec_constant<typename std::remove_reference_t<decltype(s)>::type, s>&);

}  // namespace sycl
```

This call is only valid whitin a SYCL kernel context and the `spec_constant` value must have been
created by the handler that invoked the kernel from which the `sycl::get_spec_constant` is called.

If the module has native support for specialization constant,
the returned value becomes a constant value available for the runtime compiler to use for optimisations.

**Note:** the value may become a constant (known numerical value), although this may only happen at runtime.
From the SYCL compiler's perspective (C++ compiler) the value is understood to be a runtime value and
for this reason it is not possible to use such value in a `constexpr`.

> ### Retrieving a spec constant placeholder

Reviewer 2: What happens for a program such as:
```c++
spec_id<int> id;
queue.submit([&](handler h) {
    spec_constant<int, id> c1 = h.set_spec_constant<int, id>(1);
    spec_constant<int, id> c2 = h.set_spec_constant<int, id>(2);
    h.single_task([=]{ 
        int v = c1.get_spec_constant();
    });
});
```
Is this a valid program? What's the value of `v`? Is it 1 if `native_spec_constant()` is false and 2 if `native_spec_constant()` is true?

Author: > Is this a valid program?

I think this should be banned, this look very odd and may induce unnecessary complexity on runtimes (whether there is native support or not). I'll clarify this.

> Is it 1 if native_spec_constant() is false and 2 if native_spec_constant() is true?

In all cases the result of `native_spec_constant()` does not affect the behavior.

Reviewer 2: What do you mean with ban? UB or require compiler error? I think the first isn't reasonable and the 2nd isn't possible.

Author: > What do you mean with ban?

Runtime exception.

> UB or require compiler error? I think the first isn't reasonable and the 2nd isn't possible.

i agree with you, I don't see why we would have an UB here and a compiler error is not always possible for this.

Reviewer: A runtime exception looks like a good approach to me too.
Note this behaviour is already stated on https://github.com/KhronosGroup/SYCL-Shared/pull/1/files#diff-dd15ce97b1565238f2c026fc84b30239R918 even with a more restrictive approach.

introduced usage of SPIR-V specialization constants in SYCL.
Specialization constants are associated with the SYCL 1.2.1 program class,
and named using a C++ type like kernels.
The program class gains a `set_specialization_constant` method, that sets a runtime
value to the specialization constant.
The program can then be build with `build_with_kernel_type` to create a
specialized kernel.

### Specialization constant types

On the new, type-safe, module approach, specialization constants are 
typically associated with SPIR-V module images but are not limited to such images.
The specialization constant type is defined as a general SYCL type as below:

```cpp
namespace sycl {

template <typename T>
class specialization_id {
private:
  // Implementation defined constructor.
  specialization_id(const specialization_id&) = delete;
  specialization_id(specialization_id&&) = delete;

public:

  using type = T;

  // Argument `Args` are forwarded to an underlying T Ctor.
  // This allow the user to setup a default value for the specialization_id instance.
  // The initialization of T must be evaluated at compile time to be valid.
  template<class... Args >

>   explicit constexpr specialization_id(Args&&...);

Reviewer 2: Why does this take a parameter pack rather than an argument of type T?

};

template <class T, specialization_id<T>& s>
class specialization_constant {
public:
  using type = T;

  // Create an empty (invalid) specialization constant.
  // Remains invalid until initialized by a call
  // to handler::set_specialization_constant or
  // to handler::get_specialization_constant
  specialization_constant();

  specialization_constant(const specialization_constant&) = default;
  specialization_constant& operator=(const specialization_constant&) = default;

  T get() const; // explicit access.
  operator T() const;  // implicit conversion.
};

}  // namespace sycl
```

`specialization_id` is used by the *SYCL implementation* to *reference* constant whose
concrete value will only be known during the application execution.
To create a new `specialization_id` in a module, the user creates a `specialization_id` object with an *automatic* or *static* storage duration in the namespace or class scope.

> The user then uses references to this object to bind the `specialization_constant` value to the *identifier*.

Reviewer: I do not think it binds the `specialization_constant` value to *the* *identifier*.
If I understand correctly it binds the `specialization_constant` value to the *identifier* and the command-group handler.
So a single specialization_id` object can be used to define different constant in different kernels.

> `specialization_id` objects must be forward declarable and cannot be moved or copied.

Reviewer: Now that I understand that you are interested in the address of the object and not the value of the object, I need to think about how it works with `extern` and multiple translation units.
The object could be moved or copied but at the end it is just another object with a different address, so it represents another `specialization_id`.
Also since you are interested by the address, why do you care about "objects must be forward declarable"?
This address mechanism is different from the type approach used to name kernels and global pipes in Intel extension. Is it really simpler?

Reviewer: After playing with this, it looks like the `specialization_id` objects needs static storage.
But by looking at the C++20 draft, I cannot see this restriction in *13.4.2 Template non-type arguments [temp.arg.nontype]*

Some experiment in https://github.com/keryell/teaching_C-plus-plus/blob/master/NTTP_ref.cpp and https://github.com/keryell/teaching_C-plus-plus/blob/master/NTTP_ref_2.cpp

The program can then be build with `build_with_kernel_type` to create a 
specialized kernel.

### Specialization constant types

On the new, type-safe, module approach, specialization constants are 

> typically associated with SPIR-V module images but are not limited to such image.

Reviewer: image → images ?

Author: Fixed, thanks

> The specialization constant type is defined as an general SYCL type as below:

Reviewer: an → a

```cpp
namespace sycl {

template <typename T>

> class spec_id {

Reviewer: Which is obviously 1.2.1 or 2020, right ?
The problem is that it is a niche concept, so it should not recycle some well known abbreviations...
What about just
```
 class specialization_constant_id {
```
instead? Since it is just to declare some id, it should probably fit on 1 line anyway.

Author: I reused the SPIR-V naming convention (`OpSpecConstant`) as in-fine this is to allow its representation in SYCL.

I'm fine to change `spec` to `specialization`. I don't really see a need, but you are in a better position to say if this makes things easier to understand.

Author: I made the change in this commit, https://github.com/KhronosGroup/SYCL-Shared/pull/1/commits/ad02463fe97933c35a10d5bccab75f53e991613b

I'm not convince we gain anything with this extra verbosity though.

Reviewer: We gain a clarification between `specification` and `specialization` confusion.
Since this is a niche use case, I do not think it will jeopardize the programmer productivity but will help the reader not familiar with SPIR-V details...

Author: ```
will help the reader not familiar with SPIR-V details
```

Maybe I should stop following SPIR-V then :)

If you feel it is better, we'll stick to it, no strong opinion on this

private:
  // Implementation defined constructor.
  spec_id(const spec_id&) = delete;
  spec_id(spec_id&&) = delete;

public:

  using type = T;

  // Argument `Args` are forwarded to an underlying T Ctor.
  // This allow the user to setup a default value for the spec_id instance.
  // The initialization of T must be evaluated at compile time to be valid.
  template<class... Args >
  explicit constexpr spec_id(Args&&...);
};

template <class T, spec_id<T>& s>

> class spec_constant {

Reviewer: Some possible clarifications
```
 class special_constant {
 class specialization_constant {
 class s14n_constant { :-)
 class s14n_c8t { ;-)
```

public:
  using type = T;

>   // Create an empty (invalid) spec constant.

Reviewer: ```suggestion
  // Create an empty (invalid) specialization constant.
```

  // Remains invalid until initialized by a call
  // to handler::set_spec_constant or
  // to handler::get_spec_constant
  spec_constant();

  spec_constant(const spec_constant&) = default;
  spec_constant& operator=(const spec_constant&) = default;

  T get() const; // explicit access.
  operator T() const;  // implicit conversion.
};

}  // namespace sycl
```

`spec_id` is used by the *SYCL implementation* to *reference* constant whose
concrete value will only be known during the application execution.
To create a new `spec_id` in a module, the user creates a `spec_id` object with an *automatic* or *static* storage duration in the namespace or class scope.
The user then uses references to this object to bind the `spec_constant` value to the *identifier*.

> `spec_id` objects must be forward declare-able and cannot be moved or copied.

Reviewer: What is the reason for not allowing moving or copying?

Author: A `spec_id` instance is an identifier. The sole purpose of instances of this class is to bind kernel constants with the SYCL runtime so it can manipulate it.

what would be the meaning of moving/copy an identifier ?

Reviewer: What is the rationale for *not* moving/copy an identifier ?
A typical OpenCL id in C is copyable or moveable for example.

Reviewer: OK, now I realize that it is not like a typical OpenCL id. It is just the address of the object you are using.

`spec_constant` is a *placeholder type* used by the *SYCL implementation* to
inject the real value of the specialization constant at runtime.

> For the module associated with the host device, and possibly any device with non native spec constant support, the value is carried by the `spec_constant` object it-self.

Reviewer: ```suggestion
For the module associated with the host device, and possibly any device with non native specialization constant support, the value is carried by the `spec_constant` object itself.
```

For modules natively supporting specialization constant, implementations must guaranty the values returned by the usage of the spec_constant materialized into a constant.

> ### Setting the value of spec constant

Reviewer: ```suggestion
### Setting the value of specialization constant
```

> To modify the value of a spec constant, the following API is available.

Reviewer: ```suggestion
To modify the value of a specialization constant, the following API is available.
```

> Given a SYCL module, a spec constant can be set by the `sycl::set_spec_constant` function.

Reviewer: ```suggestion
Given a SYCL module, a specialization constant can be set by the `sycl::set_spec_constant` function.
```

```cpp
namespace sycl {

template<class T, spec_id<T>& s>

> void set_spec_constant(module<module_status::input> syclModule, T);

Reviewer: If you can accept a copy of a `module`, probably `const &` would work. But I have the feeling in a real application it would be just `&`.

Author: you are right, I initially got myself mixing up with the reference counting...

It should be `&` as the function is suppose to modify the state of the module.

}  // namespace sycl
```
If C++17 or later is enabled, then the following overload is available
```cpp
namespace sycl {

template<auto& s>
void set_spec_constant(module<module_status::input> syclModule,
                       typename std::remove_reference_t<decltype(s)>::type);

}  // namespace sycl
```

> It is valid for a user to set several times a spec constant with the same `spec_id`,

Reviewer: ```suggestion
It is valid for a user to set several times a specialization constant with the same `spec_id`,
```

only the last version of the value will be consider when building the module.
Calling this function is not allowed during the submission of a task.

### Getting the value of spec constant in kernel

> To retrive the value of a spec constant in a SYCL kernel context, the following API is available.

Reviewer: ```suggestion
To retrieve the value of a specialization constant in a SYCL kernel context, the following API is available.
```

get_spec_constant(const spec_constant<typename std::remove_reference_t<decltype(s)>::type, s>&);

}  // namespace sycl
```

This call is only valid whitin a SYCL kernel context and the `spec_constant` value must have been
created by the handler that invoked the kernel from which the `sycl::get_spec_constant` is called.

If the module has native support for specialization constant,
the returned value becomes a constant value available for the runtime compiler to use for optimisations.

**Note:** the value may become a constant (known numerical value), although this may only happen at runtime.
From the SYCL compiler's perspective (C++ compiler) the value is understood to be a runtime value and
for this reason it is not possible to use such value in a `constexpr`.

### Retrieving a spec constant placeholder

A `spec_constant` handler can be retrieved from the command group handler by either:
 - Getting the placeholder for a previously set specialization constant value;
   - If no specialization constant where ever set, then the *placeholder* object holds the default value of its `spec_id`.
 - Setting a specialization constant value and retrieving the placeholder

>    - Performance notice: this function may cause the SYCL runtime to implicitly recompile the relevant SYCL module with the given value. This may have an important effect on performances and users should check their vendor recommendations on how to efficiently uses this functionality.

Reviewer: Just think FPGA... ;-) But on the other hand an implementation could cache all the previous instances too... So it really depends on the use case.

Author: You can say you don't support the feature, or rely on a similar strategy as the host one.

I added this note it may not be obvious what is happening under the hood. I expect implementations to cache generated kernels, but if the user changes the value too often, there is nothing the runtime can really do and this will create a significant slowdown.

As this function is the most convenient one, I think it is important to mention it.

    get_spec_constant();
// ...
};

}  // namespace sycl
```

| Member function                                                             | Description                                                                                                                               |
|-----------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------|
| `template<class T, spec_id<T>& s> spec_constant<T, s> set_spec_constant(T)` | Set the given value as the specialization constant value in the underlying module and returns a `spec_constant` handler usable in kernel. If the user specify to the handle a specific module to use, then the value given to `set_spec_constant` must match the the one used to build the module. In case of mismatch, the SYCL runtime raises an error. |
| `template<class T, spec_id<T>& s> spec_constant<T, s> get_spec_constant()`  | Returns a `spec_constant` handler usable in kernel.                                                                                       |

> Only one call to `set_spec_constant` or `get_spec_constant` is allowed per spec_id and per command group scope.

Reviewer: ```suggestion
Only one call to `set_spec_constant` or `get_spec_constant` is allowed per `spec_id` and per command group scope.
```

> If more than one call per spec_id and per command group scope is made then an runtime exception is thrown.

Reviewer: ```suggestion
If more than one call per `spec_id` and per command group scope is made then an runtime exception is thrown.
```

constant as a normal kernel parameter. If choosing (1), the original SPIR-V
image should be part of the SYCL module so that the module can be specialized
for non-known values.

> ### Example

Reviewer: What about starting this proposal with this example, to clarify the exposure?

The following code is showcasing how specialization constant can be used.
The code perform a simple convolution using a scaled convolution kernel.
The specialization constants are here used to inject, at runtime,
the coefficients of the kernel.

```cpp
#include <CL/sycl.hpp>

using namespace cl::sycl;

using coeff_t = std::array<std::array<float, 3>, 3>;

// Read coefficients from somewhere.
coeff_t get_coefficients();

// Identify the specialization constant.
spec_id<coeff_t> coeff_id;

void do_conv(buffer<float, 2> in, buffer<float, 2> out) {
  queue myQueue;

  myQueue.submit([&](handler &cgh) {
    auto in_acc = in.get_access<access::mode::read>(cgh);
    auto out_acc = out.get_access<access::mode::write>(cgh);

    // Create a specialization constant placeholder.

>     spec_constant<coeff_t, coeff_id> coeff =

Reviewer: DRY
```
     auto coeff = cgh.set_spec_constant<coeff_t, coeff_id>(get_coefficients());
```

Author: I wrote it this way for more clarity for someone that doesn't know anything about this, but fair enough.

        cgh.set_spec_constant<coeff_t, coeff_id>(get_coefficients());

    cgh.parallel_for<class Convolution>(
        in.get_range(), [=](cl::sycl::item<2> item_id) {
          float acc = 0;

>           for (int i = -1; i < 2; i++) {

Reviewer: Clearer from a mathematical point of view:
```
           for (int i = -1; i <= 1; i++) {
```

Author: I applied the change

>             if (item_id[0] + i < 0 || item_id[0] + i > in_acc.get_range()[0])

Reviewer: ```suggestion
            if (item_id[0] + i < 0 || item_id[0] + i >= in_acc.get_range()[0])
```

              continue;

>             for (int j = -1; j < 2; j++) {

Reviewer: ```suggestion
            for (int j = -1; j <= 1; j++) {
```

>               if (item_id[1] + j < 0 || item_id[1] + j > in_acc.get_range()[1])

Reviewer: ```suggestion
              if (item_id[1] + j < 0 || item_id[1] + j >= in_acc.get_range()[1])
```
